### ISSUE TITLE
java.lang.RuntimeException when creating native-image for Micronaut app in Docker with GraalVM 19.3.0

### ISSUE THREAD

**Issue by ilopmar** [oracle/graal#1891]
This issue is similar to https://github.com/oracle/graal/issues/1877.

There is an error when trying to build a native image for Micronaut inside Docker using `oracle/graalvm-ce:19.3.0-java8`.

Steps to reproduce the issue:

- `git clone https://github.com/micronaut-graal-tests/micronaut-basic-app`
- `cd micronaut-basic-app`
- `git checkout docker-error`
- Use Java8
- `./gradlew assemble`
- `./docker-build.sh`

The docker image creation fails with:

```
ivan@nobita:~/workspaces/micronaut-graal-tests/micronaut-basic-app $ ./docker-build.sh 
Sending build context to Docker daemon  191.2MB
Step 1/9 : FROM oracle/graalvm-ce:19.3.0-java8 as graalvm
 ---> d625000893c9
Step 2/9 : COPY . /home/app/basic-app
 ---> 0b82bb754e30
Step 3/9 : WORKDIR /home/app/basic-app
 ---> Running in 1d35893c4e78
Removing intermediate container 1d35893c4e78
 ---> 3415617fae77
Step 4/9 : RUN gu install native-image
 ---> Running in 8815085a2d8c
Downloading: Component catalog from www.graalvm.org
Processing Component: Native Image
Downloading: Component native-image: Native Image  from github.com
Installing new component: Native Image (org.graalvm.native-image, version 19.3.0)
Refreshed alternative links in /usr/bin/
Removing intermediate container 8815085a2d8c
 ---> 771a37415861
Step 5/9 : RUN native-image --no-server --static -cp build/libs/basic-app-*-all.jar
 ---> Running in 51c1457d686c
[basic-app:22]    classlist:   6,644.53 ms
[basic-app:22]        (cap):     841.08 ms
[basic-app:22]        setup:   2,403.79 ms
[basic-app:22]   (typeflow):  32,187.54 ms
[basic-app:22]    (objects):  17,779.33 ms
[basic-app:22]   (features):   2,777.07 ms
[basic-app:22]     analysis:  55,619.10 ms
[basic-app:22]     (clinit):   1,184.45 ms
[basic-app:22]     universe:   2,998.90 ms
[basic-app:22]      (parse):   4,850.88 ms
[basic-app:22]     (inline):  11,028.59 ms
[basic-app:22]    (compile):  46,348.48 ms
[basic-app:22]      compile:  64,765.04 ms
[basic-app:22]        image:   3,719.03 ms
[basic-app:22]        write:     522.38 ms
Fatal error: java.lang.RuntimeException: java.lang.RuntimeException: host C compiler or linker does not seem to work: java.lang.RuntimeException: returned 1

Running command: cc -v -o /home/app/basic-app/basic-app -z noexecstack -Wl,--gc-sections -Wl,-x -static -L/tmp/SVM-2782550808483256200 -L/opt/graalvm-ce-java8-19.3.0/jre/lib -L/opt/graalvm-ce-java8-19.3.0/jre/lib/svm/clibraries/linux-amd64 /tmp/SVM-2782550808483256200/basic-app.o /opt/graalvm-ce-java8-19.3.0/jre/lib/svm/clibraries/linux-amd64/libffi.a /opt/graalvm-ce-java8-19.3.0/jre/lib/svm/clibraries/linux-amd64/liblibchelper.a /opt/graalvm-ce-java8-19.3.0/jre/lib/svm/clibraries/linux-amd64/libjvm.a /opt/graalvm-ce-java8-19.3.0/jre/lib/svm/clibraries/linux-amd64/libstrictmath.a /opt/graalvm-ce-java8-19.3.0/jre/lib/libnet.a /opt/graalvm-ce-java8-19.3.0/jre/lib/libzip.a /opt/graalvm-ce-java8-19.3.0/jre/lib/libjava.a /opt/graalvm-ce-java8-19.3.0/jre/lib/libnio.a /opt/graalvm-ce-java8-19.3.0/jre/lib/libsunec.a /opt/graalvm-ce-java8-19.3.0/jre/lib/libsunec.a -lm -lpthread -ldl -lpthread -lz -lrt -lstdc++

Using built-in specs.
COLLECT_GCC=cc
COLLECT_LTO_WRAPPER=/usr/libexec/gcc/x86_64-redhat-linux/4.8.5/lto-wrapper
Target: x86_64-redhat-linux
Configured with: ../configure --prefix=/usr --mandir=/usr/share/man --infodir=/usr/share/info --with-bugurl=http://bugzilla.redhat.com/bugzilla --enable-bootstrap --enable-shared --enable-threads=posix --enable-checking=release --with-system-zlib --enable-__cxa_atexit --disable-libunwind-exceptions --enable-gnu-unique-object --enable-linker-build-id --with-linker-hash-style=gnu --enable-languages=c,c++,objc,obj-c++,java,fortran,ada,go,lto --enable-plugin --enable-initfini-array --disable-libgcj --with-isl=/builddir/build/BUILD/gcc-4.8.5-20150702/obj-x86_64-redhat-linux/isl-install --with-cloog=/builddir/build/BUILD/gcc-4.8.5-20150702/obj-x86_64-redhat-linux/cloog-install --enable-gnu-indirect-function --with-tune=generic --with-arch_32=x86-64 --build=x86_64-redhat-linux
Thread model: posix
gcc version 4.8.5 20150623 (Red Hat 4.8.5-39.0.1) (GCC) 
COMPILER_PATH=/usr/libexec/gcc/x86_64-redhat-linux/4.8.5/:/usr/libexec/gcc/x86_64-redhat-linux/4.8.5/:/usr/libexec/gcc/x86_64-redhat-linux/:/usr/lib/gcc/x86_64-redhat-linux/4.8.5/:/usr/lib/gcc/x86_64-redhat-linux/
LIBRARY_PATH=/usr/lib/gcc/x86_64-redhat-linux/4.8.5/:/usr/lib/gcc/x86_64-redhat-linux/4.8.5/../../../../lib64/:/lib/../lib64/:/usr/lib/../lib64/:/usr/lib/gcc/x86_64-redhat-linux/4.8.5/../../../:/lib/:/usr/lib/
COLLECT_GCC_OPTIONS='-v' '-o' '/home/app/basic-app/basic-app' '-z' 'noexecstack' '-static' '-L/tmp/SVM-2782550808483256200' '-L/opt/graalvm-ce-java8-19.3.0/jre/lib' '-L/opt/graalvm-ce-java8-19.3.0/jre/lib/svm/clibraries/linux-amd64' '-mtune=generic' '-march=x86-64'
 /usr/libexec/gcc/x86_64-redhat-linux/4.8.5/collect2 --build-id --no-add-needed --hash-style=gnu -m elf_x86_64 -static -o /home/app/basic-app/basic-app -z noexecstack /usr/lib/gcc/x86_64-redhat-linux/4.8.5/../../../../lib64/crt1.o /usr/lib/gcc/x86_64-redhat-linux/4.8.5/../../../../lib64/crti.o /usr/lib/gcc/x86_64-redhat-linux/4.8.5/crtbeginT.o -L/tmp/SVM-2782550808483256200 -L/opt/graalvm-ce-java8-19.3.0/jre/lib -L/opt/graalvm-ce-java8-19.3.0/jre/lib/svm/clibraries/linux-amd64 -L/usr/lib/gcc/x86_64-redhat-linux/4.8.5 -L/usr/lib/gcc/x86_64-redhat-linux/4.8.5/../../../../lib64 -L/lib/../lib64 -L/usr/lib/../lib64 -L/usr/lib/gcc/x86_64-redhat-linux/4.8.5/../../.. --gc-sections -x /tmp/SVM-2782550808483256200/basic-app.o /opt/graalvm-ce-java8-19.3.0/jre/lib/svm/clibraries/linux-amd64/libffi.a /opt/graalvm-ce-java8-19.3.0/jre/lib/svm/clibraries/linux-amd64/liblibchelper.a /opt/graalvm-ce-java8-19.3.0/jre/lib/svm/clibraries/linux-amd64/libjvm.a /opt/graalvm-ce-java8-19.3.0/jre/lib/svm/clibraries/linux-amd64/libstrictmath.a /opt/graalvm-ce-java8-19.3.0/jre/lib/libnet.a /opt/graalvm-ce-java8-19.3.0/jre/lib/libzip.a /opt/graalvm-ce-java8-19.3.0/jre/lib/libjava.a /opt/graalvm-ce-java8-19.3.0/jre/lib/libnio.a /opt/graalvm-ce-java8-19.3.0/jre/lib/libsunec.a /opt/graalvm-ce-java8-19.3.0/jre/lib/libsunec.a -lm -lpthread -ldl -lpthread -lz -lrt -lstdc++ --start-group -lgcc -lgcc_eh -lc --end-group /usr/lib/gcc/x86_64-redhat-linux/4.8.5/crtend.o /usr/lib/gcc/x86_64-redhat-linux/4.8.5/../../../../lib64/crtn.o
/tmp/SVM-2782550808483256200/basic-app.o:(.data+0x208): warning: Using 'dlopen' in statically linked applications requires at runtime the shared libraries from the glibc version used for linking
/usr/bin/ld: cannot find -lstdc++
/opt/graalvm-ce-java8-19.3.0/jre/lib/libnio.a(UnixNativeDispatcher.o): In function `Java_sun_nio_fs_UnixNativeDispatcher_getgrgid':
/opt/jprt/T/P1/225159.buildslave/s/jdk/src/solaris/native/sun/nio/fs/UnixNativeDispatcher.c:1037: warning: Using 'getgrgid_r' in statically linked applications requires at runtime the shared libraries from the glibc version used for linking
/opt/graalvm-ce-java8-19.3.0/jre/lib/libnio.a(UnixNativeDispatcher.o): In function `prepAttributes':
/opt/jprt/T/P1/225159.buildslave/s/jdk/src/solaris/native/sun/nio/fs/UnixNativeDispatcher.c:444: warning: Using 'getgrnam_r' in statically linked applications requires at runtime the shared libraries from the glibc version used for linking
/tmp/SVM-2782550808483256200/basic-app.o:(.data+0xb08): warning: Using 'getpwuid' in statically linked applications requires at runtime the shared libraries from the glibc version used for linking
/opt/graalvm-ce-java8-19.3.0/jre/lib/libnio.a(UnixNativeDispatcher.o): In function `prepAttributes':
/opt/jprt/T/P1/225159.buildslave/s/jdk/src/solaris/native/sun/nio/fs/UnixNativeDispatcher.c:443: warning: Using 'getpwnam_r' in statically linked applications requires at runtime the shared libraries from the glibc version used for linking
/opt/jprt/T/P1/225159.buildslave/s/jdk/src/solaris/native/sun/nio/fs/UnixNativeDispatcher.c:443: warning: Using 'getpwuid_r' in statically linked applications requires at runtime the shared libraries from the glibc version used for linking
/opt/graalvm-ce-java8-19.3.0/jre/lib/libnet.a(Inet4AddressImpl.o): In function `Java_java_net_Inet4AddressImpl_lookupAllHostAddr':
/opt/jprt/T/P1/225159.buildslave/s/jdk/src/solaris/native/java/net/Inet4AddressImpl.c:418: warning: Using 'getaddrinfo' in statically linked applications requires at runtime the shared libraries from the glibc version used for linking
collect2: error: ld returned 1 exit status

	at sun.reflect.NativeConstructorAccessorImpl.newInstance0(Native Method)
	at sun.reflect.NativeConstructorAccessorImpl.newInstance(NativeConstructorAccessorImpl.java:62)
	at sun.reflect.DelegatingConstructorAccessorImpl.newInstance(DelegatingConstructorAccessorImpl.java:45)
	at java.lang.reflect.Constructor.newInstance(Constructor.java:423)
	at java.util.concurrent.ForkJoinTask.getThrowableException(ForkJoinTask.java:593)
	at java.util.concurrent.ForkJoinTask.get(ForkJoinTask.java:1005)
	at com.oracle.svm.hosted.NativeImageGenerator.run(NativeImageGenerator.java:462)
	at com.oracle.svm.hosted.NativeImageGeneratorRunner.buildImage(NativeImageGeneratorRunner.java:315)
	at com.oracle.svm.hosted.NativeImageGeneratorRunner.build(NativeImageGeneratorRunner.java:454)
	at com.oracle.svm.hosted.NativeImageGeneratorRunner.main(NativeImageGeneratorRunner.java:115)
Caused by: java.lang.RuntimeException: host C compiler or linker does not seem to work: java.lang.RuntimeException: returned 1

Running command: cc -v -o /home/app/basic-app/basic-app -z noexecstack -Wl,--gc-sections -Wl,-x -static -L/tmp/SVM-2782550808483256200 -L/opt/graalvm-ce-java8-19.3.0/jre/lib -L/opt/graalvm-ce-java8-19.3.0/jre/lib/svm/clibraries/linux-amd64 /tmp/SVM-2782550808483256200/basic-app.o /opt/graalvm-ce-java8-19.3.0/jre/lib/svm/clibraries/linux-amd64/libffi.a /opt/graalvm-ce-java8-19.3.0/jre/lib/svm/clibraries/linux-amd64/liblibchelper.a /opt/graalvm-ce-java8-19.3.0/jre/lib/svm/clibraries/linux-amd64/libjvm.a /opt/graalvm-ce-java8-19.3.0/jre/lib/svm/clibraries/linux-amd64/libstrictmath.a /opt/graalvm-ce-java8-19.3.0/jre/lib/libnet.a /opt/graalvm-ce-java8-19.3.0/jre/lib/libzip.a /opt/graalvm-ce-java8-19.3.0/jre/lib/libjava.a /opt/graalvm-ce-java8-19.3.0/jre/lib/libnio.a /opt/graalvm-ce-java8-19.3.0/jre/lib/libsunec.a /opt/graalvm-ce-java8-19.3.0/jre/lib/libsunec.a -lm -lpthread -ldl -lpthread -lz -lrt -lstdc++

Using built-in specs.
COLLECT_GCC=cc
COLLECT_LTO_WRAPPER=/usr/libexec/gcc/x86_64-redhat-linux/4.8.5/lto-wrapper
Target: x86_64-redhat-linux
Configured with: ../configure --prefix=/usr --mandir=/usr/share/man --infodir=/usr/share/info --with-bugurl=http://bugzilla.redhat.com/bugzilla --enable-bootstrap --enable-shared --enable-threads=posix --enable-checking=release --with-system-zlib --enable-__cxa_atexit --disable-libunwind-exceptions --enable-gnu-unique-object --enable-linker-build-id --with-linker-hash-style=gnu --enable-languages=c,c++,objc,obj-c++,java,fortran,ada,go,lto --enable-plugin --enable-initfini-array --disable-libgcj --with-isl=/builddir/build/BUILD/gcc-4.8.5-20150702/obj-x86_64-redhat-linux/isl-install --with-cloog=/builddir/build/BUILD/gcc-4.8.5-20150702/obj-x86_64-redhat-linux/cloog-install --enable-gnu-indirect-function --with-tune=generic --with-arch_32=x86-64 --build=x86_64-redhat-linux
Thread model: posix
gcc version 4.8.5 20150623 (Red Hat 4.8.5-39.0.1) (GCC) 
COMPILER_PATH=/usr/libexec/gcc/x86_64-redhat-linux/4.8.5/:/usr/libexec/gcc/x86_64-redhat-linux/4.8.5/:/usr/libexec/gcc/x86_64-redhat-linux/:/usr/lib/gcc/x86_64-redhat-linux/4.8.5/:/usr/lib/gcc/x86_64-redhat-linux/
LIBRARY_PATH=/usr/lib/gcc/x86_64-redhat-linux/4.8.5/:/usr/lib/gcc/x86_64-redhat-linux/4.8.5/../../../../lib64/:/lib/../lib64/:/usr/lib/../lib64/:/usr/lib/gcc/x86_64-redhat-linux/4.8.5/../../../:/lib/:/usr/lib/
COLLECT_GCC_OPTIONS='-v' '-o' '/home/app/basic-app/basic-app' '-z' 'noexecstack' '-static' '-L/tmp/SVM-2782550808483256200' '-L/opt/graalvm-ce-java8-19.3.0/jre/lib' '-L/opt/graalvm-ce-java8-19.3.0/jre/lib/svm/clibraries/linux-amd64' '-mtune=generic' '-march=x86-64'
 /usr/libexec/gcc/x86_64-redhat-linux/4.8.5/collect2 --build-id --no-add-needed --hash-style=gnu -m elf_x86_64 -static -o /home/app/basic-app/basic-app -z noexecstack /usr/lib/gcc/x86_64-redhat-linux/4.8.5/../../../../lib64/crt1.o /usr/lib/gcc/x86_64-redhat-linux/4.8.5/../../../../lib64/crti.o /usr/lib/gcc/x86_64-redhat-linux/4.8.5/crtbeginT.o -L/tmp/SVM-2782550808483256200 -L/opt/graalvm-ce-java8-19.3.0/jre/lib -L/opt/graalvm-ce-java8-19.3.0/jre/lib/svm/clibraries/linux-amd64 -L/usr/lib/gcc/x86_64-redhat-linux/4.8.5 -L/usr/lib/gcc/x86_64-redhat-linux/4.8.5/../../../../lib64 -L/lib/../lib64 -L/usr/lib/../lib64 -L/usr/lib/gcc/x86_64-redhat-linux/4.8.5/../../.. --gc-sections -x /tmp/SVM-2782550808483256200/basic-app.o /opt/graalvm-ce-java8-19.3.0/jre/lib/svm/clibraries/linux-amd64/libffi.a /opt/graalvm-ce-java8-19.3.0/jre/lib/svm/clibraries/linux-amd64/liblibchelper.a /opt/graalvm-ce-java8-19.3.0/jre/lib/svm/clibraries/linux-amd64/libjvm.a /opt/graalvm-ce-java8-19.3.0/jre/lib/svm/clibraries/linux-amd64/libstrictmath.a /opt/graalvm-ce-java8-19.3.0/jre/lib/libnet.a /opt/graalvm-ce-java8-19.3.0/jre/lib/libzip.a /opt/graalvm-ce-java8-19.3.0/jre/lib/libjava.a /opt/graalvm-ce-java8-19.3.0/jre/lib/libnio.a /opt/graalvm-ce-java8-19.3.0/jre/lib/libsunec.a /opt/graalvm-ce-java8-19.3.0/jre/lib/libsunec.a -lm -lpthread -ldl -lpthread -lz -lrt -lstdc++ --start-group -lgcc -lgcc_eh -lc --end-group /usr/lib/gcc/x86_64-redhat-linux/4.8.5/crtend.o /usr/lib/gcc/x86_64-redhat-linux/4.8.5/../../../../lib64/crtn.o
/tmp/SVM-2782550808483256200/basic-app.o:(.data+0x208): warning: Using 'dlopen' in statically linked applications requires at runtime the shared libraries from the glibc version used for linking
/usr/bin/ld: cannot find -lstdc++
/opt/graalvm-ce-java8-19.3.0/jre/lib/libnio.a(UnixNativeDispatcher.o): In function `Java_sun_nio_fs_UnixNativeDispatcher_getgrgid':
/opt/jprt/T/P1/225159.buildslave/s/jdk/src/solaris/native/sun/nio/fs/UnixNativeDispatcher.c:1037: warning: Using 'getgrgid_r' in statically linked applications requires at runtime the shared libraries from the glibc version used for linking
/opt/graalvm-ce-java8-19.3.0/jre/lib/libnio.a(UnixNativeDispatcher.o): In function `prepAttributes':
/opt/jprt/T/P1/225159.buildslave/s/jdk/src/solaris/native/sun/nio/fs/UnixNativeDispatcher.c:444: warning: Using 'getgrnam_r' in statically linked applications requires at runtime the shared libraries from the glibc version used for linking
/tmp/SVM-2782550808483256200/basic-app.o:(.data+0xb08): warning: Using 'getpwuid' in statically linked applications requires at runtime the shared libraries from the glibc version used for linking
/opt/graalvm-ce-java8-19.3.0/jre/lib/libnio.a(UnixNativeDispatcher.o): In function `prepAttributes':
/opt/jprt/T/P1/225159.buildslave/s/jdk/src/solaris/native/sun/nio/fs/UnixNativeDispatcher.c:443: warning: Using 'getpwnam_r' in statically linked applications requires at runtime the shared libraries from the glibc version used for linking
/opt/jprt/T/P1/225159.buildslave/s/jdk/src/solaris/native/sun/nio/fs/UnixNativeDispatcher.c:443: warning: Using 'getpwuid_r' in statically linked applications requires at runtime the shared libraries from the glibc version used for linking
/opt/graalvm-ce-java8-19.3.0/jre/lib/libnet.a(Inet4AddressImpl.o): In function `Java_java_net_Inet4AddressImpl_lookupAllHostAddr':
/opt/jprt/T/P1/225159.buildslave/s/jdk/src/solaris/native/java/net/Inet4AddressImpl.c:418: warning: Using 'getaddrinfo' in statically linked applications requires at runtime the shared libraries from the glibc version used for linking
collect2: error: ld returned 1 exit status

	at com.oracle.svm.hosted.image.NativeBootImageViaCC.write(NativeBootImageViaCC.java:382)
	at com.oracle.svm.hosted.NativeImageGenerator.doRun(NativeImageGenerator.java:652)
	at com.oracle.svm.hosted.NativeImageGenerator.lambda$run$0(NativeImageGenerator.java:445)
	at java.util.concurrent.ForkJoinTask$AdaptedRunnableAction.exec(ForkJoinTask.java:1386)
	at java.util.concurrent.ForkJoinTask.doExec(ForkJoinTask.java:289)
	at java.util.concurrent.ForkJoinPool$WorkQueue.runTask(ForkJoinPool.java:1056)
	at java.util.concurrent.ForkJoinPool.runWorker(ForkJoinPool.java:1692)
	at java.util.concurrent.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:157)
Error: Image build request failed with exit status 1
The command '/bin/sh -c native-image --no-server --static -cp build/libs/basic-app-*-all.jar' returned a non-zero code: 1

```

I've also tried not using the flag `--static`. In that case I'm able to build the docker image but when running it using `frolvlad/alpine-glibc` docker image it fails with:

```
$ docker run -p 8080:8080 basic-app
/app/basic-app: error while loading shared libraries: libstdc++.so.6: cannot open shared object file: No such file or directory
```

**Comment by mzachh** [comment]
Hi Ivan,

thanks for your report!

If you add the `libstdc++-static` package to the docker image it should work like:

```
FROM oracle/graalvm-ce:19.3.0-java8 as graalvm
COPY . /home/app/basic-app
WORKDIR /home/app/basic-app
RUN gu install native-image
RUN yum install -y libstdc++-static
RUN native-image --no-server --static -cp build/libs/basic-app-*-all.jar

FROM frolvlad/alpine-glibc
EXPOSE 8080
COPY --from=graalvm /home/app/basic-app/basic-app /app/basic-app
ENTRYPOINT ["/app/basic-app"]
```

We plan to include this package in the next images.

**Comment by ilopmar** [comment]
Hi Manuel,

Thank you very much, it works :-)

**Comment by mzachh** [comment]
The package is now also included in the oracle/graalvm-ce:19.3.0* docker images.

**Comment by ilopmar** [comment]
Thank you very much!
